kafka/src/main/java/org/jboss/aerogear/unifiedpush/kafka/streams/NotificationRouterStreamsHook.java

```diff
@@ -0,0 +1,198 @@
+ * Copyright Red Hat, Inc., and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * 	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.aerogear.unifiedpush.kafka.streams;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import javax.ejb.Singleton;
+import javax.ejb.Startup;
+import javax.enterprise.concurrent.ManagedExecutorService;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import net.wessendorf.kafka.serialization.CafdiSerdes;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+
+import org.jboss.aerogear.unifiedpush.api.FlatPushMessageInformation;
+import org.jboss.aerogear.unifiedpush.api.PushApplication;
+import org.jboss.aerogear.unifiedpush.api.Variant;
+import org.jboss.aerogear.unifiedpush.api.VariantType;
+import org.jboss.aerogear.unifiedpush.kafka.KafkaClusterConfig;
+import org.jboss.aerogear.unifiedpush.kafka.serdes.InternalUnifiedPushMessageSerde;
+import org.jboss.aerogear.unifiedpush.kafka.serdes.PushApplicationSerde;
+import org.jboss.aerogear.unifiedpush.message.InternalUnifiedPushMessage;
+import org.jboss.aerogear.unifiedpush.message.holder.MessageHolderWithVariants;
+import org.jboss.aerogear.unifiedpush.service.GenericVariantService;
+import org.jboss.aerogear.unifiedpush.service.metrics.PushMessageMetricsService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+import static org.jboss.aerogear.unifiedpush.api.VariantType.*;
+
+/**
+ * Class that will be initialized on startup, which reads messages from the input topic
+ * {@link KafkaClusterConfig#NOTIFICATION_ROUTER_STREAMS_INPUT_TOPIC} containing key/value pair (PushApplication, InternalUnifiedPushMessage).
+ *
+ * These records are split into subrecords, based on the message's variants. This allows messages to be processed separately,
+ * giving attention to limitations and requirements of specific push networks.
+ *
+ * Once processing is performed, records are streamed to output topics based on their variant type for further processing
+ * in {@link org.jboss.aerogear.unifiedpush.message.token.TokenLoader}.
+ */
+@Singleton
+@Startup
+public class NotificationRouterStreamsHook {
```

Comment:
and here, let's use CDI 

`@javax.enterprise.context.ApplicationScoped` annotation